common/src/main/java/io/netty/util/concurrent/CompletionStageAdapter.java

```diff
@@ -0,0 +1,527 @@
+        return whenCompleteAsync(action);
+    }
+
+    @Override
+    public CompletionStage<V> whenCompleteAsync(BiConsumer<? super V, ? super Throwable> action) {
+        return whenCompleteAsync(action, executor());
+    }
+
+    @Override
+    public CompletionStage<V> whenCompleteAsync(BiConsumer<? super V, ? super Throwable> action, Executor executor) {
+        Promise<V> promise = executor().newPromise();
+        future.addListener(f -> {
+            if (inEventLoop(executor)) {
+                whenCompleteAsync0(promise, f, action);
+            } else {
+                safeExecute(executor, () -> whenCompleteAsync0(promise, f, action), promise);
+            }
+        });
+        return promise.asStage();
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <U, V> void whenCompleteAsync0(
+            Promise<U> promise, Future<? super V> f, BiConsumer<? super V, ? super Throwable> action) {
+        try {
+            action.accept((V) f.getNow(), f.cause());
+        } catch (Throwable cause) {
+            promise.setFailure(cause);
+            return;
+        }
+        promise.setSuccess(null);
+    }
+
+    @Override
+    public <U> CompletionStage<U> handle(BiFunction<? super V, Throwable, ? extends U> fn) {
+        return handleAsync(fn);
+    }
+
+    @Override
+    public <U> CompletionStage<U> handleAsync(BiFunction<? super V, Throwable, ? extends U> fn) {
+        return handleAsync(fn, executor());
+    }
+
+    @Override
+    public <U> CompletionStage<U> handleAsync(BiFunction<? super V, Throwable, ? extends U> fn, Executor executor) {
+        Promise<U> promise = executor().newPromise();
+        future.addListener(f -> {
+            if (inEventLoop(executor)) {
+                handleAsync0(promise, f, fn);
+            } else {
+                safeExecute(executor, () -> handleAsync0(promise, f, fn), promise);
+            }
+        });
+        return promise.asStage();
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <U, V> void handleAsync0(
+            Promise<U> promise, Future<? super V> f, BiFunction<? super V, Throwable, ? extends U> fn) {
+        final U result;
+        try {
+            result = fn.apply((V) f.getNow(), f.cause());
+        } catch (Throwable cause) {
+            promise.setFailure(cause);
+            return;
+        }
+        promise.setSuccess(result);
+    }
+
+    private static boolean inEventLoop(Executor executor) {
+        return executor instanceof EventExecutor && ((EventExecutor) executor).inEventLoop();
+    }
+
+    private static void safeExecute(Executor executor, Runnable task, Promise<?> promise) {
+        try {
+            executor.execute(task);
+        } catch (Throwable cause) {
+            promise.setFailure(cause);
+        }
+    }
+
+    @Override
+    public CompletableFuture<V> toCompletableFuture() {
+        throw new UnsupportedOperationException();
```

Comment:
Nit: Add a descriptive error reason. Assume this will end up in some poor soul's application log and they have no idea why.

Also consider documenting this as unsupported in the class Javadoc.

Reply:
done...